File: src/domain/reconcile.ts
import { addMoney, subtractMoney, type Halalas } from './money'
import type { IsoDate } from './entities/types'

/**
 * مطابقة الرصيد الجاري — spec/02 و OVERRIDES §7-ب.
 *
 * القاعدة: افتتاحي + الوارد − الصادر، **بترتيب المصدر**.
 * spec/02: «لا تفترض ترتيب عمليات اليوم إذا غاب دليل» —
 * لذلك يُحفظ ترتيب السطر في الملف (sourceOrder) ويُستخدم كما هو،
 * ولا يُعاد الترتيب داخل اليوم.
 */

/** حركة واحدة على المحفظة بترتيب المصدر. */
export interface LedgerMovement {
  date: IsoDate
  /** ترتيب السطر في الملف الأصلي — يحفظ ترتيب اليوم كما ورد. */
  sourceOrder: number
  debitMinor: Halalas // خارج
  creditMinor: Halalas // داخل
  /** الرصيد المعلن في الكشف بعد هذه الحركة، إن وُجد. */
  statedBalanceMinor?: Halalas
  /** مرجع للعرض عند اكتشاف فرق. */
  reference?: string
  label?: string
}

export interface BalanceMismatch {
  index: number
  date: IsoDate
  computedMinor: Halalas
  statedMinor: Halalas
  differenceMinor: Halalas
  reference?: string
  label?: string
  /** هل يشترك هذا اليوم في أكثر من حركة؟ لو نعم فالغموض في الترتيب وارد. */
  sameDayCount: number
}

export interface ReconcileResult {
  openingMinor: Halalas
  openingAt: IsoDate
  closingMinor: Halalas
  closingAt: IsoDate | null
  totalDebitMinor: Halalas
  totalCreditMinor: Halalas
  movementCount: number
  /** الحركات التي خالف فيها المحسوب المعلن. فارغة = مطابقة كاملة. */
  mismatches: BalanceMismatch[]
  /** عدد الحركات التي فيها رصيد معلن، أي القابلة للمقارنة أصلًا. */
  checkedCount: number
  /**
   * تفسير الغموض عند وجود فروق في يوم متعدد الحركات.
   * spec/06: «توضيح الغموض؛ لا ادعاء تحديد أول فرق يقينًا».
   */
  ambiguityNote: string | null
}

/**
 * يطبّق الحركات بالترتيب على رصيد افتتاحي، ويقارن بالرصيد المعلن حيث وُجد.
 *
 * لا يُعاد ترتيب المدخل — الترتيب المُمرَّر هو ترتيب المصدر.
 */
export function reconcileBalance(
  openingMinor: Halalas,
  openingAt: IsoDate,
  movements: readonly LedgerMovement[],
): ReconcileResult {
  const perDay = new Map<IsoDate, number>()
  for (const m of movements) perDay.set(m.date, (perDay.get(m.date) ?? 0) + 1)

  let balance = openingMinor
  let totalDebit = 0
  let totalCredit = 0
  let checked = 0
  const mismatches: BalanceMismatch[] = []

  movements.forEach((m, index) => {
    totalDebit = addMoney(totalDebit, m.debitMinor)
    totalCredit = addMoney(totalCredit, m.creditMinor)
    balance = addMoney(subtractMoney(balance, m.debitMinor), m.creditMinor)

    if (m.statedBalanceMinor === undefined) return
    checked++
    if (balance !== m.statedBalanceMinor) {
      mismatches.push({
        index,
        date: m.date,
        computedMinor: balance,
        statedMinor: m.statedBalanceMinor,
        differenceMinor: subtractMoney(balance, m.statedBalanceMinor),
        reference: m.reference,
        label: m.label,
        sameDayCount: perDay.get(m.date) ?? 1,
      })
    }
  })

  let ambiguityNote: string | null = null
  if (mismatches.length > 0) {
    const first = mismatches[0]
    ambiguityNote =
      first.sameDayCount > 1
        ? `أول فرق في ${first.date}، واليوم ده فيه ${first.sameDayCount} حركة. ` +
          `ترتيب حركات اليوم الواحد مش مثبت في الكشف، فمش أكيد إن دي بالذات هي أول عملية فيها مشكلة.`
        : `أول فرق في ${first.date}، واليوم ده فيه حركة واحدة، فالفرق يخصها هي.`
  }

  return {
    openingMinor,
    openingAt,
    closingMinor: balance,
    closingAt: movements.length > 0 ? movements[movements.length - 1].date : null,
    totalDebitMinor: totalDebit,
    totalCreditMinor: totalCredit,
    movementCount: movements.length,
    mismatches,
    checkedCount: checked,
    ambiguityNote,
  }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Reconcile.kt
package app.masroufy.core

/**
 * مطابقة الرصيد الجاري — نقل `src/domain/reconcile.ts` (spec/02 و OVERRIDES §7-ب).
 * افتتاحي + الوارد − الصادر **بترتيب المصدر** — مفيش إعادة ترتيب جوه اليوم.
 */
data class LedgerMovement(
    val date: IsoDate,
    /** ترتيب السطر في الملف الأصلي. */
    val sourceOrder: Int,
    val debitMinor: Halalas,
    val creditMinor: Halalas,
    /** الرصيد المعلن بعد الحركة، لو موجود. */
    val statedBalanceMinor: Halalas? = null,
    val reference: String? = null,
    val label: String? = null,
)

data class BalanceMismatch(
    val index: Int,
    val date: IsoDate,
    val computedMinor: Halalas,
    val statedMinor: Halalas,
    val differenceMinor: Halalas,
    val reference: String?,
    val label: String?,
    /** اليوم فيه كام حركة؟ أكتر من واحدة = الترتيب ممكن يكون هو السبب. */
    val sameDayCount: Int,
)

data class ReconcileResult(
    val openingMinor: Halalas,
    val openingAt: IsoDate,
    val closingMinor: Halalas,
    val closingAt: IsoDate?,
    val totalDebitMinor: Halalas,
    val totalCreditMinor: Halalas,
    val movementCount: Int,
    /** فاضية = مطابقة كاملة. */
    val mismatches: List<BalanceMismatch>,
    /** الحركات اللي فيها رصيد معلن (القابلة للمقارنة). */
    val checkedCount: Int,
    /** «توضيح الغموض؛ لا ادعاء تحديد أول فرق يقينًا» (spec/06). */
    val ambiguityNote: String?,
)

fun reconcileBalance(openingMinor: Halalas, openingAt: IsoDate, movements: List<LedgerMovement>): ReconcileResult {
    val perDay = movements.groupingBy { it.date }.eachCount()
    var balance = openingMinor
    var totalDebit = 0L
    var totalCredit = 0L
    var checked = 0
    val mismatches = mutableListOf<BalanceMismatch>()
    movements.forEachIndexed { index, m ->
        totalDebit = addMoney(totalDebit, m.debitMinor)
        totalCredit = addMoney(totalCredit, m.creditMinor)
        balance = addMoney(subtractMoney(balance, m.debitMinor), m.creditMinor)
        val stated = m.statedBalanceMinor ?: return@forEachIndexed
        checked++
        if (balance != stated) {
            mismatches += BalanceMismatch(index, m.date, balance, stated, subtractMoney(balance, stated), m.reference, m.label, perDay[m.date] ?: 1)
        }
    }
    val first = mismatches.firstOrNull()
    val note = when {
        first == null -> null
        first.sameDayCount > 1 ->
            uiText(TextKey.RECONCILE_FIRST_GAP_MANY, first.date, first.sameDayCount.toString())
        else -> uiText(TextKey.RECONCILE_FIRST_GAP_ONE, first.date)
    }
    return ReconcileResult(
        openingMinor, openingAt, balance, movements.lastOrNull()?.date, totalDebit, totalCredit,
        movements.size, mismatches, checked, note,
    )
}
